File: data/src/main/java/com/arturogutierrez/openticator/storage/realm/helpers/RealmObservable.java
package com.arturogutierrez.openticator.storage.realm.helpers;

import io.realm.Realm;
import io.realm.RealmObject;
import io.realm.RealmResults;
import rx.Observable;
import rx.functions.Func1;

public final class RealmObservable {

  private RealmObservable() {

  }

  public static <T extends RealmObject> Observable<T> object(final Func1<Realm, T> function) {
    return Observable.create(new OnSubscribeRealm<T>() {
      @Override
      public T get(Realm realm) {
        return function.call(realm);
      }
    });
  }

  public static <T extends RealmObject> Observable<RealmResults<T>> results(
      final Func1<Realm, RealmResults<T>> function) {
    return Observable.create(new OnSubscribeRealmResults<T>() {
      @Override
      public RealmResults<T> get(Realm realm) {
        return function.call(realm);
      }
    });
  }
}

File: data/src/main/kotlin/com/arturogutierrez/openticator/storage/realm/helpers/RealmObservable.kt
package com.arturogutierrez.openticator.storage.realm.helpers

import io.realm.Realm
import io.realm.RealmObject
import io.realm.RealmResults
import rx.Observable
import rx.functions.Func1

object RealmObservable {

    fun <T : RealmObject> `object`(function: Func1<Realm, T>): Observable<T> {
        return Observable.create(object : OnSubscribeRealm<T>() {
            override fun get(realm: Realm): T {
                return function.call(realm)
            }
        })
    }

    fun <T : RealmObject> results(
            function: Func1<Realm, RealmResults<T>>): Observable<RealmResults<T>> {
        return Observable.create(object : OnSubscribeRealmResults<T>() {
            override fun get(realm: Realm): RealmResults<T> {
                return function.call(realm)
            }
        })
    }
}
